packages/button/src/vaadin-button-base.js

```diff
@@ -3,70 +3,86 @@
  * Copyright (c) 2017 - 2025 Vaadin Ltd.
  * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
  */
+import '@vaadin/component-base/src/style-props.js';
 import { css } from 'lit';
 
 export const buttonStyles = css`
-  :host {
-    display: inline-block;
-    position: relative;
-    outline: none;
-    white-space: nowrap;
-    -webkit-user-select: none;
-    user-select: none;
-  }
+  @layer base {
+    :host {
+      display: inline-flex;
+      align-items: center;
+      justify-content: center;
+      text-align: center;
+      gap: var(--vaadin-button-gap, 0 var(--_vaadin-gap-icon));
 
-  :host([hidden]) {
-    display: none !important;
-  }
+      white-space: nowrap;
+      -webkit-tap-highlight-color: transparent;
+      -webkit-user-select: none;
+      user-select: none;
+      cursor: pointer;
+      box-sizing: border-box;
+      vertical-align: middle;
+      flex-shrink: 0;
+      height: var(--vaadin-button-height, auto);
+      margin: var(--vaadin-button-margin, 0);
+      padding: var(--vaadin-button-padding, var(--_vaadin-padding-container-text));
+      font-family: var(--vaadin-button-font-family, inherit);
+      font-size: var(--vaadin-button-font-size, inherit);
+      line-height: var(--vaadin-button-line-height, inherit);
+      font-weight: var(--vaadin-button-font-weight, 500);
+      color: var(--vaadin-button-text-color, var(--_vaadin-color-strong));
+      background: var(--vaadin-button-background, var(--_vaadin-bg-container));
+      background-origin: border-box;
+      border: var(
+        --vaadin-button-border,
+        var(--vaadin-button-border-width, 1px) solid var(--vaadin-button-border-color, var(--_vaadin-border-color))
+      );
+      border-radius: var(--vaadin-button-border-radius, var(--_vaadin-radius-m));
+    }
 
-  :host([disabled]) {
-    pointer-events: var(--_vaadin-button-disabled-pointer-events, none);
-    cursor: not-allowed;
-  }
+    :host([hidden]) {
+      display: none !important;
+    }
 
-  /* Aligns the button with form fields when placed on the same line.
-  Note, to make it work, the form fields should have the same "::before" pseudo-element. */
-  .vaadin-button-container::before {
-    content: '\\2003';
-    display: inline-block;
-    width: 0;
-    max-height: 100%;
-  }
+    .vaadin-button-container,
+    [part='prefix'],
+    [part='suffix'],
+    [part='label'] {
+      display: contents;
+    }
 
-  .vaadin-button-container {
-    display: inline-flex;
-    align-items: center;
-    justify-content: center;
-    text-align: center;
-    width: 100%;
-    height: 100%;
-    min-height: inherit;
-    text-shadow: inherit;
-  }
+    :host([focus-ring]) {
+      outline: var(--vaadin-focus-ring-width) solid var(--vaadin-focus-ring-color);
```

Comment:
What do you think about button specific focus ring variables that can be used? For example we had the accessibility requirement that a primary button has a different focus-ring.. allowing me to explicity style only the focus rings of buttons with `--vaadin-button-focus-ring-color` which falls back to `--vaadin-focus-ring-color` if not set would make this easily feasible.

Reply:
I wouldn't add those in the base styles at least. If you want to customize the focus outline for buttons, I would say you write the selector and override the properties:

```css
vaadin-button[theme~="primary"] {
  --vaadin-focus-ring-color: ...;
}
```

I’m curious, what was the specific accessibility requirement you had for primary buttons?

Reply:
Our accessibility tester complained that the blue focus ring of the blue button does not really convey that it's focused - we had to change it to another color. You can see something similar also on github:

Cancel Focused: 

![grafik](https://github.com/user-attachments/assets/7c2e67c1-6238-4777-b3b7-0b00958e13f9)

Comment Focused:

![grafik](https://github.com/user-attachments/assets/9c1e6fbb-59b2-49aa-a486-5137416798f7)

Reply:
Thanks, that makes sense. I believe that a 1px or 2px offset/gap between the focus outline and the button would also suffice as a solution. That’s what I opted to do for the button base style now, that there’s always a 1px outline-offset.